Sanitize innert document

IgorMinar opened the pull request.

mhevery committed:
refactor($sanitize): new implementation of the html sanitized parser

This implementation is based on using inert document parsed by the browser

Closes #11442
Closes #11443

IgorMinar committed:
fix($sanitize): support void elements, fixups, remove dead code, typos

IgorMinar committed:
feat($sanitize): make svg support an opt-in

BREAKING CHANGE: The svg support in  is now an opt-in option

Applications that depend on this option can use  to turn the option back on,
but while doing so, please read the warning provided in the  documentation for
information on preventing click-hijacking attacks when this option is turned on.

IgorMinar commented:
hmm.. tests are failing because different browsers enumerate over the list of tags in a different order which results in non-deterministic. I'll either have to sort the attribute list or change the matcher to account for this.

lgalfaso commented:
@IgorMinar even after sorting the attributes, there are two test failures with IE9

```
IE 9.0.0 (Windows 7) HTML htmlParser should throw badparse if text content contains "<" followed by an ASCII letter without matching ">" FAILED
	Expected { tag : 'a', attrs : { body : '', bar< : '' } } to equal undefined.
IE 9.0.0 (Windows 7) HTML htmlParser should parse empty value attribute of node FAILED
	Expected null to equal { tag : 'option', attrs : { selected : '', value : '' } }.
```

I think it should be possible to fix the second issue with something like https://github.com/angular/angular.js/blob/master/src/jqLite.js#L159
The first issue, I am not sure on how to work around it

IgorMinar committed:
fix($compile): properly sanitize xlink:href attribute interoplation

IgorMinar committed:
chore(travis): disable browserstack builds for now

we don't really pay attention to them anyway.

gkalpak commented on src/ngSanitize/sanitize.js:
```diff
@@ -153,6 +154,42 @@ function $SanitizeProvider() {
+   * Enables a subset of svg to be supported by the sanitizer.
+   *
+   * **Warning**: By enabling this setting without taking other precautions, you might expose your
+   * application to click-hijacking attacks. In these attacks, a sanitize svg could be positioned
```
_sanitize svg_ ??

IgorMinar replied:
fixed.

gkalpak commented on src/ngSanitize/sanitize.js:
```diff
@@ -153,6 +154,42 @@ function $SanitizeProvider() {
+   *
+   * @param {boolean=} regexp New regexp to whitelist urls with.
+   * @returns {boolean|ng.$sanitizeProvider} Returns the currently configured value if called
+   *    without a argument or self for chaining otherwise.
```
_a_  -->  _an_

IgorMinar replied:
fixed. thanks.

hackvertor commented:
Found a bug with closing br elements. Closing </br> get converted to opening br's
</br> -> <br>

IgorMinar committed:
fix($sanitize): add mXSS protection

petebacondarwin commented on .travis.yml:
```diff
@@ -19,10 +19,10 @@ env:
+#    - JOB=unit BROWSER_PROVIDER=browserstack
+#    - JOB=docs-e2e BROWSER_PROVIDER=browserstack
+#    - JOB=e2e TEST_TARGET=jqlite BROWSER_PROVIDER=browserstack
+#    - JOB=e2e TEST_TARGET=jquery BROWSER_PROVIDER=browserstack
```
Is this disabling just temporary or should we give up with browserstack altogether?

IgorMinar replied:
I think it's just temporary

petebacondarwin commented on src/ngSanitize/sanitize.js:
```diff
@@ -153,6 +154,42 @@ function $SanitizeProvider() {
+   * @description
+   * Enables a subset of svg to be supported by the sanitizer.
+   *
+   * **Warning**: By enabling this setting without taking other precautions, you might expose your
```
I would put this warning inside a `<div class="alert alert-warning">` block to make it stand out more.

petebacondarwin commented on src/ngSanitize/sanitize.js:
```diff
@@ -262,11 +289,30 @@ function makeMap(str, lowercaseKeys) {
+  }
+  var docElement = doc.documentElement || doc.getDocumentElement();
+  var bodyElements = docElement.getElementsByTagName('body');
+  if (bodyElements.length === 1) {
```
Is it possible that `bodyElements` can have `length > 1`? If so, can we not just take the first one?

IgorMinar replied:
usually there should be only one body, but in IE there is none, so we need to create it. I'll add a comment to that code

petebacondarwin commented on src/ngSanitize/sanitize.js:
```diff
@@ -276,169 +322,76 @@ function makeMap(str, lowercaseKeys) {
+        break;
+    }
+    var nextNode;
+    if (!(nextNode = node.firstChild)) {
```
I think this block can be simplified:

```js
    var nextNode;
    if (!(nextNode = node.firstChild)) {
      if (node.nodeType == 1) {
        handler.end(node.nodeName.toLowerCase());
      }
      nextNode = node.nextSibling;
      if (!nextNode) {
        while (nextNode == null) {
          node = node.parentNode;
          if (node === inertBodyElement) break;
          nextNode = node.nextSibling;
          if (node.nodeType == 1) {
            handler.end(node.nodeName.toLowerCase());
          }
        }
      }
    }
    node = nextNode;
```

IgorMinar replied:
nice catch

petebacondarwin commented:
A few small comments but otherwise LGTM

petebacondarwin commented:
How far back can this be backported?

IgorMinar commented:
@petebacondarwin since there are breaking changes included in here, I don't think that we should backport this to older versions. It's safer to keep it in master only.

IgorMinar committed:
fix($sanitize): strip urls starting with 'unsafe:' as opposed to 'unsafe'

IgorMinar committed:
refactor($sanitize): remove <script> from valid block elements

the script and style tag are explicitly blacklisted, so this doesn't change any functionality.
the change is done to improve code clarity.

IgorMinar committed:
refactor(): rename local variables to improve code clarity

IgorMinar committed:
test($sanitize): add a test to prove that html comments are being stripped

h3rald commented:
@IgorMinar, @petebacondarwin 

Hi, what do you mean with "it's safer to keep it in master only"? Is there any plan to include this security fix in a 1.4.x release?

petebacondarwin commented:
> safer to keep it in master only

This means that since there are breaking changes, trying to backport it to older versions could cause havoc with live applications. The message is that if you are concerned about these security features then you should upgrade to 1.5 as soon as it comes out.

h3rald commented:
Hi Pete, thanks a lot for the clarification.

I was wondering if you or someone else could elaborate a bit more on the vulnerability fixed by this pull request: in particular, is it browser-dependent? Does the vulnerability apply just to some/older browsers? Which?

Again thanks a lot for your time!

petebacondarwin commented:
I'll have a further chat with @IgorMinar about this.

IgorMinar commented:
aside from the option to disable svg support (which is not a security issue if you set `overflow: hidden` on the element containing the sanitized output), I don't believe that there is a reason for considering the previous implementation insecure. The new implementation is simpler and easier to maintain going forward.

The reason why we didn't backport this to 1.4 was that the new implementation could parse html slightly differently which could result in hard-to-debug issues in the rendered output  (but not security vulnerabilities).

h3rald commented:
Hi Igor,

Thanks a lot for your comment: it really clarifies things a lot!

My original concern about backporting was due to the fact that this particular pull request was notified to me as a "fix" for possible vulnerabilities affecting Angular sanitization. Based on your comments (and also after doing a bit of testing with both implementations at this point) I think that perhaps the security notification I received was perhaps a bit misleading or simply tagged incorrectly. I will investigate further within my company on why this happened. 

Thanks a lot for your support!

moshere commented:
> Hi Pete, thanks a lot for the clarification.
> 
> I read through the commits and the only breaking change seems to be the fact that SVG sanitization is not enabled by default (and if enabled you recommend extra care), right?
> 
> The fact that you are now relying on the browser itself to parse HTML code by creating an inert HTML document should not be too much of a problem because AFAIK this functionality is available in all major browsers (including IE9+...
> 
> I was notified of this fix by a security notification issued throughout my company which recommended that we apply the patch manually as a temporary workaround until the fix is officially distributed.
> If the fix is not applied, apparently it may be possible to perform a XSS attack in certain conditions -- is it right?
> 
> Thanks again for your time!
> 
> P.S.: I actually tried to apply the fixes to v1.4.3 and it doesn't seem to break much (I still have to test it with directives using SVG though).

@h3rald - Can I get the relevant fix for 1.4.x version? thank you!
